### DISCUSSION TITLE
RXBIN_PATH as include directory

### DISCUSSION THREAD

**Discussion by mgrossmann** [adesutherland/CREXX#202] [Ideas]
Hello all,

in the course of setting up my development environment and thinking about feature f0024, 
I had the idea, again analogous to Java, to introduce an environment variable RXBIN_PATH. 
At this point is searched for dependent *.rxbin files.

Later, of course, the "internal" bifs should be handled separately.

What do you think?

**Comment by adesutherland** [comment]
Perhaps RXLIB_PATH?

**Comment by mgrossmann** [comment]
What about both?

RXLIB_PATH for kind of linked library of all functions

RXBIN_PATH for the single rxbin files?

or RXLIB_PATH for both?

Java have the CLASS_PATH for single class files
an a java.library.path for C libraries.

But in the end, it‘s just a name….

Should I create a feature for this?
Yea, should I branch from develop?

Another question, should I add a cmake target
for building all rexx and asm functions?

Yes, I’m the same branch or in the f0024 branch?


A lot of questions, I know. 🤓

Mike

Adrian Sutherland ***@***.***> schrieb am Do. 11. Nov. 2021
um 12:00:

> Perhaps RXLIB_PATH?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/adesutherland/CREXX/discussions/202#discussioncomment-1625077>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAF7Q7IOAFA37B5ONJN6EM3ULOO5VANCNFSM5H2DSCTQ>
> .
>
-- 
Von Gmail Mobile gesendet

**Comment by adesutherland** [comment]
> What about both?

My preference is something like CREXX_PATH for everything (REXX, RXAS, RXBIN, RXLIB) to make it super simple? But if you think this will actually cause issues in real life then please feel free to split it up! You are right users may want different paths for different things ... 

> should I branch from develop?

Yes please make a feature f0024 from develop for rxas, rxas etc

> should I add a cmake target for building all rexx and asm functions?

Yes, but that needs to be done in f0035, I think

**Reply by mgrossmann** [reply]
sounds good...

**Comment by mgrossmann** [comment]
another question;

When creating a build lib target, where should I put the generated rxbin files?
I found a `native` sub directory in `lib/rxfns`. Is this ment to hold the rxbin files or is it for native implementations of bifs?
As we will not check in the assembled files, I could create a temp directory. e.g. `lib/rxfns/lib_build`.

Mike

**Reply by rvjansen** [reply]
@mgrossmann This is indeed meant for native implementation. The BIF is one, admittedly special, type of library, but I would like to keep this structure for other libraries that might have native components. I foresee subdirectories with hw instruction set architecture, os, and os release.

**Comment by adesutherland** [comment]
> Is this ment to hold the rxbin files or is it for native implementations of bifs?

This is for native implementations of bifs ...

Should the built rxbin targets go into the CMAKE target directory structure under lib/rexx etc. i.e. the same way that c built stuff goes?

**Reply by mgrossmann** [reply]
That would be the cleanest option, yes.

But I see two things then. First, why should we separate rexx and rxas? In the end, rxbin files are generated from all functions (from rexx and rxas). Maybe `cmake-build-target/lib/rxfns` would be the right directory then? The second question is, how does `rxvm` find these generated rxbin files. But then the new environment variable would come into play. CMake should then automatically export an environment variable that points to `cmake-build-target/lib/rxfns`, right?

**Reply by adesutherland** [reply]
> First, why should we separate rexx and rxas?

We have competing implementations of the same BiF at the moment. Perhaps we can retire some of them, @rvjansen?

Perhaps CMAKE should be configured to put RXC, RXAS, RXVM etc. and all the RXBINs into **one** target directory (which would work well for users) and RX should always search first in the directory of the binary. Also when you have rxar you can put all the rxbins into a single rxlib file, perhaps called runtime.rxlib which is automatically included by rxln and/or rxvm.

**Reply by rvjansen** [reply]
We should, in my opinion, not separate them according to initial source type. Rexx BIF's too, become rxas.
Secondly, we should retire only after rigoruous, repeatable performance testing. Also, when we understand better what happens - the rxas substr is faster in a standalone test, and slower in the loop. The other way around (Rexx<rxas) or (loop>1 time) is also imaginable. So I would like to be conservative here - we choose what we package anyway.
Third, in my NetRexx career I have experienced that every environment variable is a stumbling block for users. I think in the bog standard distribution everything should be statically linked, including the standard BIF library. Of course, for *us* it is very desirable to have paths and libpaths, but let's keep them optional for users and usable for power users.

**Reply by rvjansen** [reply]
@mgrossmann First, why should we separate rexx and rxas?
This is done because some .rxas is expendable, and some .rxas is precious. Separating the source directories is the safest option. We envisage a selection process, steered by functional tests and performance tests, in which we choose one which is bound into a library. For ourselves we need the mechanism to run with different libraries (including standard); we need a mechanism in which users can override our choice in favour of their own implementation - of which they have the source - and the choice of running with other libraries that they only have in binary form. And with native binaries e.g. from other packages.

**Reply by rvjansen** [reply]
Which all does not mean the we should not have a search path now - indeed it is a bit tedious to look up the dependencies for every function. Maybe RXBIN for individual .rxbin and RXLIB for libraries, even. In which we specify that RXBIN has precedence.

**Reply by mgrossmann** [reply]
a lot of input ...

let my brain do some work, now. :)

The original question was because I had to do a lot of compiles and assembles, before I could start my first rexx.
So let us do step by step and iterate over these problems.

**Reply by rvjansen** [reply]
I've got a makefile that does its thing. As we don't have depencies, it works everywhere:
```
#
# makefile for crexx programs

.mxas.rxas:
	cpp -E $*.mxas | awk '/^[^#]/{ print }' > $*.rxas
	rxas $*
#	rxdas $*

.rexx.rxas:
	rxc $*

.rxas.rxbin:
	rxas $*
#	rxdas $*


RX_SRC		:= $(wildcard *.rexx)
AS_MXA		:= $(wildcard *.mxas)
AS_RXA		:= $(wildcard *.rxas)
AS_SRC		:= $(AS_RXA:.rxas=.rxbin)
MAS_SRC		:= $(AS_MXA:.mxas=.rxas)
RX_BIN		:= $(RX_SRC:.rexx=.rxbin)

.SUFFIXES: .rexx .rxas .mxas .rxbin

#
# target all compiles
#
all::	 $(MAS_SRC) $(AS_SRC) $(RX_BIN)

.PRECIOUS: %.rxas

#
# target clean removes compiled products
#
.PHONY: clean
clean:
	rm -f *.rxbin
	find . -name "*.rexx" | awk '{$$2 = $$1 ; sub ( /\.rexx/,".rxas",$$1 ) ; print $$1 }' | xargs rm -f
	find . -name "*.mxas" | awk '{$$2 = $$1 ; sub ( /\.mxas/,".rxas",$$1 ) ; print $$1 }' | xargs rm -f
```

**Reply by mgrossmann** [reply]
Hi @rvjansen 

could you please take a look at my last commit on f0035?
I added kind of build target for the rxas and rexx subfolder.
I'm not sure, how this compares with your make file.

Another thing, is it wanted, that some compiled test functions are commited to the lib/rxfns/rxas folder?

PS: to be able to make the project on windows, I have changed your MV commands in the test sub folder.

Best, Mike

**Reply by rvjansen** [reply]
Hi Mike, works!
I think the rule should be, we don't commit .rxbin. We also don't commit .rxas if we have .rexx, we only do that when we only have .rxas. The test suite should just build and run. So for the test functions the same rule applies. We do have some test functions in .rxas at the moment; they should be in the lib/rxfns/rxas subdirectory, not in tests_functional or tests_performance.

**Reply by mgrossmann** [reply]
Hi René,

> I think the rule should be, we don't commit .rxbin. We also don't commit .rxas if we have .rexx, we only do that when we only have .rxas. 

I agree.

> We do have some test functions in .rxas at the moment; they should be in the lib/rxfns/rxas subdirectory, not in tests_functional or tests_performance.

Why? Because currently the corresponding rexx files have their home in tests_functional.
(I'm just asking for interest.)

**Reply by rvjansen** [reply]
Two reasons actually - they were there first, but the second is more important - they will not be at risk of being erased on a make clean. This when building in-source, I understand we build out-of-source in the project, but I never saw any reason to, when there is only one executable, like .class, or .rxbin here. Now when building native executables for instruction set architectures, os'es and releases, that is another thing of course. But I would like to be able to build in-source for these things and do the occasional make clean.

**Reply by rvjansen** [reply]
But, reading your message better, there should not be any tests .rxbin's that are in the other folders. Will have a look at that.

**Reply by rvjansen** [reply]
[I have deleted one remark about the makefile, I did some local damage but that appears to be my own fault.]

**Comment by rvjansen** [comment]
Hi Mike, I noticed something strange when playing around with kernel probes. On my test system, which has Regina, brexx, oorexx, NetRexx and Pierre Richard's Jaxx available, rxvm was not minding its own business while having brexx's string library on LD_LIBRARY_PATH. See:
```
227900 rxvm               -1   2 /usr/local/lib/tls/haswell/x86_64/libm.so.6
227900 rxvm               -1   2 /usr/local/lib/tls/haswell/libm.so.6
227900 rxvm               -1   2 /usr/local/lib/tls/x86_64/libm.so.6
227900 rxvm               -1   2 /usr/local/lib/tls/libm.so.6
227900 rxvm               -1   2 /usr/local/lib/haswell/x86_64/libm.so.6
227900 rxvm               -1   2 /usr/local/lib/haswell/libm.so.6
227900 rxvm               -1   2 /usr/local/lib/x86_64/libm.so.6
227900 rxvm               -1   2 /usr/local/lib/libm.so.6
227900 rxvm               -1   2 /home/rvjansen/apps/brexx/lstring/tls/haswell/x86_64/libm.so.6
227900 rxvm               -1   2 /home/rvjansen/apps/brexx/lstring/tls/haswell/libm.so.6
227900 rxvm               -1   2 /home/rvjansen/apps/brexx/lstring/tls/x86_64/libm.so.6
227900 rxvm               -1   2 /home/rvjansen/apps/brexx/lstring/tls/libm.so.6
227900 rxvm               -1   2 /home/rvjansen/apps/brexx/lstring/haswell/x86_64/libm.so.6
227900 rxvm               -1   2 /home/rvjansen/apps/brexx/lstring/haswell/libm.so.6
227900 rxvm               -1   2 /home/rvjansen/apps/brexx/lstring/x86_64/libm.so.6
227900 rxvm               -1   2 /home/rvjansen/apps/brexx/lstring/libm.so.6
227900 rxvm                3   0 /etc/ld.so.cache
227900 rxvm                3   0 /lib/x86_64-linux-gnu/libm.so.6
227900 rxvm               -1   2 /usr/local/lib/libc.so.6
227900 rxvm               -1   2 /home/rvjansen/apps/brexx/lstring/libc.so.6
227900 rxvm                3   0 /lib/x86_64-linux-gnu/libc.so.6
227900 rxvm                3   0 ./test1.rxbin
227900 rxvm                3   0 ./../rexx/length.rxbin
227900 rxvm                3   0 ./../rxas/substr.rxbin
227900 rxvm                3   0 ./../rexx/raise.rxbin
227900 rxvm                3   0 ./../rxas/right.rxbin
227900 rxvm                3   0 ./../rxas/left.rxbin
227900 rxvm                3   0 ./../rxas/copies.rxbin
227900 rxvm                3   0 ./../rexx/trunc.rxbin
```
When emptying LD_LIBRARY_PATH, this becomes:
```
228016 rxvm               -1   2 /usr/local/lib/tls/haswell/x86_64/libm.so.6
228016 rxvm               -1   2 /usr/local/lib/tls/haswell/libm.so.6
228016 rxvm               -1   2 /usr/local/lib/tls/x86_64/libm.so.6
228016 rxvm               -1   2 /usr/local/lib/tls/libm.so.6
228016 rxvm               -1   2 /usr/local/lib/haswell/x86_64/libm.so.6
228016 rxvm               -1   2 /usr/local/lib/haswell/libm.so.6
228016 rxvm               -1   2 /usr/local/lib/x86_64/libm.so.6
228016 rxvm               -1   2 /usr/local/lib/libm.so.6
228016 rxvm                3   0 /etc/ld.so.cache
228016 rxvm                3   0 /lib/x86_64-linux-gnu/libm.so.6
228016 rxvm               -1   2 /usr/local/lib/libc.so.6
228016 rxvm                3   0 /lib/x86_64-linux-gnu/libc.so.6
228016 rxvm                3   0 ./test1.rxbin
228016 rxvm                3   0 ./../rexx/length.rxbin
228016 rxvm                3   0 ./../rxas/substr.rxbin
228016 rxvm                3   0 ./../rexx/raise.rxbin
228016 rxvm                3   0 ./../rxas/right.rxbin
228016 rxvm                3   0 ./../rxas/left.rxbin
228016 rxvm                3   0 ./../rxas/copies.rxbin
228016 rxvm                3   0 ./../rexx/trunc.rxbin
```
So that might be something to take care of when implementing the search paths.

**Reply by mgrossmann** [reply]
Hi @rvjansen 

strange thing with the LD_LIBRARY_PATH....
I'm currently not quiet sure, how to interpret this.

Mike

**Comment by mgrossmann** [comment]
René,

you didn‘t read it wrong. It wasn’t written very well. With compiled files
the read files was meant.

It‘s not important, now.

But:
Someone has to take care of the list of rxas and rexx files, that will be
compiled and assembled by cmake. I just grabbed all but the test cases.
Need double checking. 🤓

Thanks,
Mike

René Vincent Jansen ***@***.***> schrieb am Mo. 15. Nov. 2021
um 15:24:

> But, reading your message better, there should not be any tests .rxbin's
> that are in the other folders. Will have a look at that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/adesutherland/CREXX/discussions/202#discussioncomment-1645009>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAF7Q7KZ22H3NINHEDT3GITUMEJ2PANCNFSM5H2DSCTQ>
> .
>
-- 
Von Gmail Mobile gesendet
